File: 2015/03/src/main.rs
use std::fs;
use std::collections::HashSet;

type Pos = (i64, i64);
type Houses = HashSet<Pos>;

fn read_file(filepath: &str) -> String {
	let contents = fs::read_to_string(filepath);
	let binding = contents.expect("REASON");
	binding.split('\n').next().unwrap().to_string()
}

fn found_houses(line: &str) -> usize {
	let mut houses: Houses = HashSet::new();
	let mut pos: Pos = (0,0);
	houses.insert(pos);
	for c in line.chars() {
		match c {
			'>' => pos = (pos.0, pos.1 + 1),
			'<' => pos = (pos.0, pos.1 - 1),
			'v' => pos = (pos.0 + 1, pos.1),
			'^' => pos = (pos.0 - 1, pos.1),
			 _  => pos = pos,
		}
		houses.insert(pos);
	}
	houses.len()
}

fn robo_santa(line: &str) -> usize {
	let mut houses: Houses = HashSet::new();
	let mut pos: Pos = (0,0);
	let mut robo_pos: Pos = (0,0);
	houses.insert(pos);
	for c in line.chars() {
		match c {
			'>' => pos = (pos.0, pos.1 + 1),
			'<' => pos = (pos.0, pos.1 - 1),
			'v' => pos = (pos.0 + 1, pos.1),
			'^' => pos = (pos.0 - 1, pos.1),
			 _  => pos = pos,
		}
		houses.insert(pos);
		(pos, robo_pos) = (robo_pos, pos);
	}
	houses.len()
}


fn part1() {
	let line = read_file("INPUT");
	println!("Part 1: {}", found_houses(&line));
}

fn part2() {
	let line = read_file("INPUT");
	println!("Part 2: {}", robo_santa(&line));
}

fn main() {
	println!("Year 2015 day 3 - Perfectly Spherical Houses in a Vacuum");
	part1();
	part2();
}

File: 2015/03/main.exs
IO.puts("Year 2015 day 3 - Perfectly Spherical Houses in a Vacuum")

if {:ok, string} = File.read("INPUT") do
	{map, _, _} = string |> String.graphemes()
	|> List.foldl({%{}, 0,0}, fn(x, {map, acc_x, acc_y}) ->
	cond do
		x == "<" -> {Map.put_new(map, "#{acc_x}X#{acc_y}", true), acc_x - 1, acc_y}
		x == ">" -> {Map.put_new(map, "#{acc_x}X#{acc_y}", true), acc_x + 1, acc_y}
		x == "^" -> {Map.put_new(map, "#{acc_x}X#{acc_y}", true), acc_x, acc_y - 1}
		x == "v" -> {Map.put_new(map, "#{acc_x}X#{acc_y}", true), acc_x, acc_y + 1}
		true     -> {Map.put_new(map, "#{acc_x}X#{acc_y}", true), acc_x, acc_y}
	end end)
	IO.puts("Part 1: #{map_size(map)}")
end

if {:ok, string} = File.read("INPUT") do
	{map, _, _, _, _, _} = string |> String.graphemes()
	|> List.foldl({%{},0,0,false, 0, 0}, fn(x, {map, acc_x, acc_y, robo, robo_x, robo_y}) ->
	cond do
		x == "<" and not robo -> {Map.put_new(map, "#{acc_x}X#{acc_y}", true), acc_x - 1, acc_y, true, robo_x, robo_y}
		x == ">" and not robo -> {Map.put_new(map, "#{acc_x}X#{acc_y}", true), acc_x + 1, acc_y, true, robo_x, robo_y}
		x == "^" and not robo -> {Map.put_new(map, "#{acc_x}X#{acc_y}", true), acc_x, acc_y - 1, true, robo_x, robo_y}
		x == "v" and not robo -> {Map.put_new(map, "#{acc_x}X#{acc_y}", true), acc_x, acc_y + 1, true, robo_x, robo_y}
		x == "<" and robo     -> {Map.put_new(map, "#{robo_x}X#{robo_y}", true), acc_x, acc_y, false, robo_x - 1, robo_y}
		x == ">" and robo     -> {Map.put_new(map, "#{robo_x}X#{robo_y}", true), acc_x, acc_y, false, robo_x + 1, robo_y}
		x == "^" and robo     -> {Map.put_new(map, "#{robo_x}X#{robo_y}", true), acc_x, acc_y, false, robo_x, robo_y - 1}
		x == "v" and robo     -> {Map.put_new(map, "#{robo_x}X#{robo_y}", true), acc_x, acc_y, false, robo_x, robo_y + 1}
		not robo              -> {Map.put_new(map, "#{acc_x}X#{acc_y}", true), acc_x, acc_y, true, robo_x, robo_y}
		robo                  -> {Map.put_new(map, "#{robo_x}X#{robo_y}", true), acc_x, acc_y, false, robo_x, robo_y}
	end end)
	IO.puts("Part 2: #{map_size(map)}")
end
